### PR TITLE
Integrate new optimization module based on linopy

### PR DESCRIPTION
**Pull request by FabianHofmann**
## Changes proposed in this Pull Request

A new sub-module "optimization" for the linopy-based optimization is introduced. The optimization module is sub-structured in 
* variables.py - for all variable definitions
* constraints.py  - for all local constraint definitions
* global_constraints.py - for  all global constraints
* common.py - for common utility functions
* optimize.py - for functions running the sanitiy checks, optimization and solution assignment
* abstract.py - for optimization functions abstracted from the normal lopf, (like iterative lopf, possibly also sclopf and others) 

**Remaining Todo's**

* [x] sanitize check
    * [X] nans in nominal bounds
    * [x] nans in operational bounds
    * [x] ~link efficiency < 1 #319~ 
    * [x] intersection of committable and extendable assets
* [x] support multi-port links
* [x] support min_up_time min_down_time constraints
* [x] make `p_set` argument available for all static and time-dependent variables #202 
* [x] **documentation**  

**Run it**

Run the optimization by 
```
n.optimize()
```
you can also create the model and optimize afterwards 
```
n.optimize.create_model()
# do whatever stuff
n.optimize.solve_model()
```
You can access variables in 
```
n.model.variables
```
and constraints in 
```
n.model.constraints
```
From there you can look up and modify the LP, like for example
```
n.model.constraints['my_constraint'].lhs = n.model.constraints['my_constraint'].lhs + another_variable
```
Happy to have some feedback @nworbmot @fneum @lisazeyen @pz-max @p-glaum 

## Preview Documentation

https://pypsa--336.org.readthedocs.build/en/336/

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added.
- [x] Newly introduced dependencies are added to `environment.yaml`, `environment_docs.yaml` and `setup.py` (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

**Comment by nworbmot** [comment]
Hi @FabianHofmann, this is very impressive! I'm really enjoying `linopy` - much better than `linopt`.
A few comments:
- Something seems to be going wrong with multi-links. The [sector-coupling example](https://pypsa.readthedocs.io/en/latest/examples/biomass-synthetic-fuels-carbon-management.html) gives different answers with `n.lopf` and `n.optimize`, and `n.optimize` is breaking on PyPSA-Eur-Sec.
- Could you give a few examples with custom constraints? E.g. is it possible to do:
```
n.create_model()
p = n.model["Link-ext-p"]
p_nom = n.model["Link-p_nom"]
lhs = linexpr((43,p),(-243,p_nom))
n.model.add_constraint(lhs,"=", 45)
n.model.solve()
```
I think something close to this is what you had in mind...

**Comment by FabianHofmann** [comment]
@nworbmot great! 

Did you check that the solvers are the same? `linopy` uses by default gurobi and the multi-link optimizations are often degenerate. At least  this is what I saw that in the [reservoir example](https://pypsa.readthedocs.io/en/latest/examples/chained-hydro-reservoirs.html).

Yes, quiet close. Could you try
```
# use the `optimize` accessor (multiple functions available)
n.optimize.create_model()

p = n.model["Link-p"]
p_nom = n.model["Link-p_nom"].rename({'Link-ext': 'Link'})

lhs = 43 * p - 243 * p_nom
n.model.add_constraints(lhs == 45) # a bit more compact

# again use the optimize accessor, this solves the model and adds the results to the network
n.optimize.solve_model()
```
One tricky thing is to have in mind that different dimension names are always broadcasted in arithmetic operations (like the xarray convention wants it). So one has to take care what dimensions the variables have. 
 

I know more examples are lacking. I don't think I can make today. But if you have specific questions, please don't hesitate.

**Comment by nworbmot** [comment]
Thanks @FabianHofmann! I meant the objective function answer is different by a factor 2, so it's not just degeneracy for the biomass example.
BTW PyPSA-Eur-Sec is failing in the `define_nodal_balance_constraints` function, which I guess will be to do with the multi-links.
Thanks for help with custom constraints, that's a very nice interface!

**Comment by FabianHofmann** [comment]
@nworbmot thanks for double-checking things. I found the bug. I did not allow for negative emissions. 
Could you send me the trace-back of the sector-coupling network error?

**Comment by nworbmot** [comment]
Perfect, now it works! Actually for the PyPSA-Eur-Sec it is giving the same objective function (hooray!) there just seem to be some bugs with assignment and issues when you modify the network inbetween. I'll send you the notebook by email.

**Comment by FabianHofmann** [comment]
@fneum @p-glaum this is ready to be merged. To keep in mind for the next conda release: add linopy as a dependency there. 
The implementation is highly tested, Unit commitment, multi-investment, security-lopf, ilopf, global constraints is all in. 

Important:
- [ ] Could you have a look at the [introduction example](https://pypsa--336.org.readthedocs.build/en/336/examples/optimization-with-linopy.html) and [migration guide](https://pypsa--336.org.readthedocs.build/en/336/examples/optimization-with-linopy-migrate-extra-functionalities.html). 
- [ ] Test it against a pypsa-eur-(sec) network

**Comment by pz-max** [comment]
@FabianHofmann really excited about this. Just noted that the introduction example has some mini issues related to "Gurobi" and probably wrong variable naming.

**Comment by FabianHofmann** [comment]
Thanks @pz-max, saw that as well, fixing it now.

File: pypsa/optimization/optimize.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+
+    @is_documented_by(iterative_transmission_capacity_expansion)
+    def iterative_transmission_capacity_expansion(self, *args, **kwargs):
+        iterative_transmission_capacity_expansion(self._parent, *args, **kwargs)
+
+    @is_documented_by(security_constraint_optimization)
+    def security_constraint_optimization(self, *args, **kwargs):
+        security_constraint_optimization(self._parent, *args, **kwargs)
+
+    def fix_optimal_capacities(self):
```
Comment (fneum) [inline pypsa/optimization/optimize.py:513]: You could also have a utility function `fix_optimal_dispatch()` that prepares a solved network for the run of `n.pf()`:

```py
n.links_t.p_set = n.links_t.p0
n.generators_t.p_set = n.generators_t.p
n.storage_units_t.p_set = n.storage_units_t.p
...
```

File: pypsa/optimization/compat.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Use compatibility methods for optimization problem definition with Linopy.
```
Comment (fneum) [inline pypsa/optimization/compat.py:4]: Nice idea to ease the transition!

File: pypsa/optimization/abstract.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    ## add costs of additional infrastructure to objective value of last iteration
+    obj_links = (
+        n.links[ext_dc_links_b].eval("capital_cost * (p_nom_opt - p_nom_min)").sum()
+    )
+    obj_lines = n.lines.eval("capital_cost * (s_nom_opt - s_nom_min)").sum()
+    n.objective += obj_links + obj_lines
+    n.objective_constant -= obj_links + obj_lines
+
+
+def security_constraint_optimization(
```
Comment (fneum) [inline pypsa/optimization/abstract.py:157]:
```suggestion
def optimize_security_constrained(
```
active formulation maybe better?

File: pypsa/optimization/abstract.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from pypsa.descriptors import nominal_attrs
+
+logger = logging.getLogger(__name__)
+
+
+def iterative_transmission_capacity_expansion(
```
[the remark this answers is not in the document]
Comment (fneum) [reply pypsa/optimization/abstract.py:19]:
```suggestion
def optimize_iteratively(
```
active formulation may be better?

**Review by fneum — COMMENTED** [review]
Really, really nice!

I had a first quick check, with a particular focus on how the new code is linked to existing code to check for possible breaking changes! This looks very clean.

Also, the new optimisation module is well tested! Any remaining quick wins to boost test coverage? Wondering why there's no Codecov report coming up.

**Comment by fneum** [comment]
Some parts of https://pypsa--336.org.readthedocs.build/en/336/optimal_power_flow.html could use an update to added optimisation module if you have time.

**Comment by fneum** [comment]
Example and migration guide are excellent!

**Comment by fneum** [comment]
> Wondering why there's no Codecov report coming up.
Problem with CI setup, fixed in `master`, code coverage should pop up soon.

File: pypsa/optimization/optimize.py (hunk 1 above)
Comment (FabianHofmann) [reply pypsa/optimization/optimize.py:513]: love that!

File: environment.yaml [hunk 5]
```diff
@@ -17,3 +17,5 @@ dependencies:
 - deprecation
 #  - coincbc
 #  - gurobi::gurobi
+- pip:
+  - linopy
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply environment.yaml:21]: Not yet...

File: test/test_lopf_unit_commitment.py [hunk 6]
```diff
@@ -153,3 +253,128 @@ def test_minimum_down_time():
     expected_dispatch = np.array([[0, 0, 3000, 8000], [3000, 800, 0, 0]], dtype=float).T
 
     equal(nu.generators_t.p.values, expected_dispatch)
+
+
+@pytest.mark.parametrize("api", ["pyomo", "linopy"])
```
[the remark this answers is not in the document]
Comment (FabianHofmann) [reply test/test_lopf_unit_commitment.py:258]: I don't think we mention it in the documentation. Only in the lopf itself (when unit commitment is activated).

**Comment by FabianHofmann** [comment]
Thanks @fneum for your feedback! Here the point I'll address:

- [ ] rename abstract optimization functions to active name
- [ ] small adjustment of the optimization section in the documenation
- [ ]  add `fix_optimal_dispatch`
- [ ] rename SUPPORTED_APIS

File: environment.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (fneum) [reply environment.yaml:21]: Not urgent :)

**Comment by FabianHofmann** [comment]
@fneum tried to retrigger the codecov bot, but don't know how... It's ready to be merged now :)

**Comment by fneum** [comment]
Great, let's do it! Today's changes all look good. :) You can do the honours!

Should we directly tag a 0.21 release?

**Comment by FabianHofmann** [comment]
Great, tomorrow I'll have a look at #333, after that let's do the release

**Comment by FabianHofmann** [comment]
Hey @toby-coleman, probably next week :)

**Pull request merged**